pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,308 @@
+}
+
+func (v *VolcanoBatchScheduler) addVolcanoAnnotation(pod *corev1.Pod, app *v1beta1.SparkApplication) *util.PatchOperation {
+	path := "/metadata/annotations"
+	var value interface{}
+	if len(pod.Annotations) != 0 {
+		if _, ok := pod.Annotations[v1alpha2.GroupNameAnnotationKey]; ok {
+			return nil
+		}
+		pod.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		value = pod.Annotations
+	} else {
+		value = map[string]string{v1alpha2.GroupNameAnnotationKey: v.getAppPodGroupName(app)}
+	}
+	return &util.PatchOperation{Op: "replace", Path: path, Value: value}
+}
+
+func (v *VolcanoBatchScheduler) BeforeSubmitSparkApplication(app *v1beta1.SparkApplication) error {
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupInClientMode(app)
+	} else if app.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupInClusterMode(app)
+	}
+	return nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClientMode(app *v1beta1.SparkApplication) error {
+	//Only executor resource will be considered.
+	return v.syncPodGroup(app, 1, getExecutorRequestResource(app))
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClusterMode(app *v1beta1.SparkApplication) error {
+	totalResource := sumResourceList([]corev1.ResourceList{getExecutorRequestResource(app), getDriverRequestResource(app)})
+	//NOTE: In cluster mode, the initial size of PodGroup is set to 1 in order to schedule driver pod first.
+	return v.syncPodGroup(app, 1, totalResource)
+}
+
+func (v *VolcanoBatchScheduler) getAppPodGroupName(app *v1beta1.SparkApplication) string {
+	return fmt.Sprintf("spark-%s-pg", app.Name)
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroup(app *v1beta1.SparkApplication, size int32, minResource corev1.ResourceList) error {
+	var err error
+	podGroupName := v.getAppPodGroupName(app)
+	if pg, err := v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Get(podGroupName, v1.GetOptions{}); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		podGroup := v1alpha2.PodGroup{
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: app.Namespace,
+				Name:      podGroupName,
+				OwnerReferences: []v1.OwnerReference{
+					*v1.NewControllerRef(app, v1beta1.SchemeGroupVersion.WithKind("SparkApplication")),
+				},
+			},
+			Spec: v1alpha2.PodGroupSpec{
+				MinMember:    size,
+				MinResources: &minResource,
+			},
+		}
+
+		_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Create(&podGroup)
+	} else {
+		if pg.Spec.MinMember != size {
+			pg.Spec.MinMember = size
+			_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Update(pg)
+		}
+	}
+	if err != nil {
+		glog.Errorf(
+			"Unable to sync PodGroup with error: %s. Abandon schedule pods via volcano.", err)
+	}
+	return err
+}
+
+func New(config *rest.Config, webhookEnabled bool) schedulerinterface.BatchScheduler {
+
+	if !webhookEnabled {
+		glog.Error("failed to initialize volcano client, webhook enable is required")
```

Comment:
`failed to initialize the volcano scheduler as it requires the webhook to be enabled`